posts/onboarding.md

```diff
@@ -0,0 +1,525 @@
+
+If a token has been consumed but could not be parsed, the parser can move backward to the previous token by calling `_retreat()` to decrement the index.
+
+In some scenarios, the parser’s behavior depends on both the current and surrounding tokens. The parser can “peek” at the previous and next tokens without consuming them by accessing the `_prev` and `_next` properties, respectively.
+
+### Matching tokens and text
+When parsing a SQL statement, the parser must identify specific sets of keywords to correctly build the AST. It does so by “matching” sets of tokens or keywords to infer the structure of the statement.
+
+For example, these matches occur when parsing the window specification `SUM(x) OVER (PARTITION BY y)`:
+1. The `SUM` and `(` tokens are matched and `x` is parsed as an identifier
+2. The `)` token is matched
+2. The `OVER`  and `(` tokens are matched
+2. The `PARTITION` and `BY` tokens are matched
+3. The partition clause `y` is parsed
+4. The `)` token is matched
+
+The family of `_match` methods perform different varieties of matching. They return `True` and advance the index if there is a match, or return `None` if the tokens do not match.
+
+The most commonly used `_match` methods are:
+
+
+Method                | Use
+-------------------------|-----
+**`_match(type)`**       | Attempt to match a single token with a specific `TokenType`
+**`_match_set(types)`**  | Attempt to match a single token in a set of `TokenType`s
+**`_match_text_seq(texts)`**   | Attempt to match a continuous sequence of strings/texts
+**`_match_texts(texts)`**   | Attempt to match a keyword in a set of strings/texts
+
+### `_parse` methods
+SQLGlot’s parser follows the “recursive descent” approach, meaning that it relies on mutually recursive methods in order to understand SQL syntax and the various precedence levels between combined SQL expressions.
+
+For example, the SQL statement `CREATE TABLE a (col1 INT)` will be parsed into an `exp.Create` expression that includes the kind of object being created (table) and its schema (table name, column names and types). The schema will be parsed into an `exp.Schema` node that contains one column definition `exp.ColumnDef` node for each column.
+
+The relationships between SQL constructs are encoded in methods whose names begin with “_parse_”, such as `_parse_create()`. We refer to these as “parsing methods”.
+
+For example, to parse the statement above SQLGlot calls the `_parse_create()` method. Then:
+- `_parse_create()` determines that a `TABLE` is being created
+- Because a table is being created, it knows the table name comes next and calls `_parse_table_parts()` (the table name may have multiple parts like `catalog.schema.table`)
+- Because a table is being created, it knows the schema definition comes after the table name and calls `_parse_schema()`
```

Comment:
```suggestion
- It also knows that a schema definition comes after the table name and calls `_parse_schema()`
```

Reply:
Hmm, I think if I was to change the description here I'd go with something like:

```
For example, to parse the statement above the entrypoint method `_parse_create()` is called. Then:
- In `_parse_create()`, the parser determines that a `TABLE` is being created
- Because a table is being created, the table name is expected next and thus `_parse_table_parts()` is invoked; This is because the table name may have multiple parts such as `catalog.schema.table`
- Having processed the table name, the parser continues parsing the schema definition by calling `_parse_schema()`
```

It may have a bit more noise added, but I think the bullet point breakdown is stitched together more "naturally". What do you say?

Reply:
Sure yeah